File: frontend/src/api/asset_types.ts
/**
 * api/asset_types.ts — 項目類型管理 API
 *
 * 項目類型（AssetType）是訂閱的分類標籤，例如「軟體授權」、「雲端服務」等。
 * 由管理員在系統設定頁面維護；訂閱表單與 Dashboard 的篩選下拉選單皆使用此清單。
 *
 * 刪除限制：後端會檢查是否有訂閱正在使用該類型，若有則拒絕刪除並回傳錯誤訊息。
 */
import { api } from './client'
import type { ApiResponse, AssetType } from '@/types/api'

/** 取得所有項目類型清單，依建立時間排序 */
export async function listAssetTypes(): Promise<AssetType[]> {
  const { data } = await api.get<ApiResponse<AssetType[]>>('/api/v1/asset-types')
  if (!data.success || !data.data) throw new Error(data.message)
  return data.data
}

/** 建立新的項目類型，名稱不可重複（後端會拒絕並回傳 409 衝突錯誤） */
export async function createAssetType(name: string): Promise<AssetType> {
  const { data } = await api.post<ApiResponse<AssetType>>('/api/v1/asset-types', { name })
  if (!data.success || !data.data) throw new Error(data.message)
  return data.data
}

/** 更新項目類型的名稱 */
export async function updateAssetType(id: number, name: string): Promise<AssetType> {
  const { data } = await api.patch<ApiResponse<AssetType>>(`/api/v1/asset-types/${id}`, { name })
  if (!data.success || !data.data) throw new Error(data.message)
  return data.data
}

/**
 * 刪除項目類型。
 * 若有訂閱正在使用此類型，後端會回傳 409 錯誤，需先將訂閱改為其他類型再刪除。
 */
export async function deleteAssetType(id: number): Promise<void> {
  await api.delete(`/api/v1/asset-types/${id}`)
}

File: backend/src/api/v1/routers/asset_types.py
from application.use_cases.create_asset_type import CreateAssetTypeUseCase
from application.use_cases.delete_asset_type import DeleteAssetTypeUseCase
from application.use_cases.list_asset_types import ListAssetTypesUseCase
from application.use_cases.update_asset_type import UpdateAssetTypeUseCase
from domain.entities.user import User
from fastapi import APIRouter, Depends
from infrastructure.database.repositories.asset_type_repository import SqlAssetTypeRepository
from infrastructure.database.session import get_db
from sqlalchemy.ext.asyncio import AsyncSession

from api.dependencies import get_current_user, require_admin
from api.v1.schemas.asset_type import AssetTypeCreate, AssetTypeResponse, AssetTypeUpdate
from api.v1.schemas.base import ApiResponse

router = APIRouter(prefix="/api/v1/asset-types", tags=["asset-types"])


def _get_repo(db: AsyncSession = Depends(get_db)) -> SqlAssetTypeRepository:
    return SqlAssetTypeRepository(db)


@router.get("", response_model=ApiResponse[list[AssetTypeResponse]])
async def list_asset_types(
    _: User = Depends(get_current_user),
    repo: SqlAssetTypeRepository = Depends(_get_repo),
) -> ApiResponse[list[AssetTypeResponse]]:
    items = await ListAssetTypesUseCase(repo).execute()
    return ApiResponse.ok(data=[AssetTypeResponse(**vars(t)) for t in items])


@router.post("", response_model=ApiResponse[AssetTypeResponse], status_code=201)
async def create_asset_type(
    body: AssetTypeCreate,
    current_user: User = Depends(get_current_user),
    repo: SqlAssetTypeRepository = Depends(_get_repo),
) -> ApiResponse[AssetTypeResponse]:
    result = await CreateAssetTypeUseCase(repo).execute(name=body.name, created_by=current_user.id)
    return ApiResponse.ok(data=AssetTypeResponse(**vars(result)))


@router.patch("/{asset_type_id}", response_model=ApiResponse[AssetTypeResponse])
async def update_asset_type(
    asset_type_id: int,
    body: AssetTypeUpdate,
    _: User = Depends(require_admin),
    repo: SqlAssetTypeRepository = Depends(_get_repo),
) -> ApiResponse[AssetTypeResponse]:
    result = await UpdateAssetTypeUseCase(repo).execute(asset_type_id=asset_type_id, name=body.name)
    return ApiResponse.ok(data=AssetTypeResponse(**vars(result)))


@router.delete("/{asset_type_id}", response_model=ApiResponse[None])
async def delete_asset_type(
    asset_type_id: int,
    _: User = Depends(require_admin),
    repo: SqlAssetTypeRepository = Depends(_get_repo),
) -> ApiResponse[None]:
    await DeleteAssetTypeUseCase(repo).execute(asset_type_id=asset_type_id)
    return ApiResponse.ok(message="已刪除")
